File: Cpp/1138_Alphabet_Board_Path.cpp
class Solution {
public:
    string alphabetBoardPath(string target) {
        string result = "";
        int prev_r = 0;
        int prev_c = 0;
        
        for(auto& ch: target) {
            int r = (ch - 'a') / 5;
            int c = (ch - 'a') % 5;
            
            int diff_r = r - prev_r;
            int diff_c = c - prev_c;
            
            if(diff_r < 0) {
                result += string(-diff_r, 'U');
            }
            
            if(diff_c < 0){
                result += string(-diff_c, 'L');
            }
            
            if(diff_r > 0){
                result += string(diff_r, 'D');
            }
            
            if(diff_c > 0){
                result += string(diff_c, 'R');
            }
            
            result += '!';
            
            
            prev_r = r;
            prev_c = c;
        }
        
        return result;
    }
};

File: Kotlin/1138_Alphabet_Board_Path.kt
class Solution {
    fun alphabetBoardPath(target: String): String {
        val result = StringBuilder()
        var prevR = 0
        var prevC = 0
        
        target.forEach{ ch ->
            val r = (ch - 'a') / 5
            val c = (ch - 'a') % 5
            
            val diffR = r - prevR
            val diffC = c - prevC
            
            repeat(-diffR) {
                result.append('U')
            }
            repeat(-diffC) {
                result.append('L')
            }
            repeat(diffR) {
                result.append('D')
            }
            repeat(diffC) {
                result.append('R')
            }
            
            result.append('!')
            
            prevR = r
            prevC = c
        }
        
        return result.toString()
    }
}